File: server-js/models/metadata.js
const pool = require('../db/db');

const metadata = async (id) => {
  try {
    const query = await pool.query(`
      SELECT (
          json_build_object(
              'product_id', id,
              'ratings', json_build_object(
                  '1', num_1_stars,
                  '2', num_2_stars,
                  '3', num_3_stars,
                  '4', num_4_stars,
                  '5', num_5_stars
              ),
              'recommended', num_recommended,
              'characteristics', json_build_object(
                  'Fit', json_build_object(
                      'id', fit_id,
                      'value', fit_total::float / num_reviews
                  ),
                  'Size', json_build_object(
                      'id', size_id,
                      'value', size_total::float / num_reviews
                  ),
                  'Width', json_build_object(
                      'id', width_id,
                      'value', width_total::float / num_reviews
                  ),
                  'Comfort', json_build_object(
                      'id', comfort_id,
                      'value', comfort_total::float / num_reviews
                  ),
                  'Quality', json_build_object(
                      'id', quality_id,
                      'value', quality_total::float / num_reviews
                  ),
                  'Length', json_build_object(
                      'id', length_id,
                      'value', length_total::float / num_reviews
                  )
              )
          )
      )
      FROM reviews.products WHERE reviews.products.id = $1;
    `, [id])

    return query.rows;
  } catch (error) {
    return error;
  }
}

module.exports = metadata;

File: server-rs/src/metadata.rs
use rocket::serde::{Deserialize, Serialize};

#[path = "./db.rs"]
mod db;

#[derive(Deserialize, Serialize)]
#[serde(crate = "rocket::serde")]
pub struct Metadata {
    pub id: i32,
    pub num_1_stars: Option<i32>,
    pub num_2_stars: Option<i32>,
    pub num_3_stars: Option<i32>,
    pub num_4_stars: Option<i32>,
    pub num_5_stars: Option<i32>,
    pub recommended: Option<i32>,
    pub characteristics: Characteristics,
}

#[derive(Deserialize, Serialize)]
#[serde(crate = "rocket::serde")]
pub struct Characteristic {
    pub id: Option<i32>,
    pub value: Option<f64>,
}

#[derive(Deserialize, Serialize)]
#[serde(crate = "rocket::serde")]
pub struct Characteristics {
    pub Fit: Characteristic,
    pub Size: Characteristic,
    pub Width: Characteristic,
    pub Comfort: Characteristic,
    pub Quality: Characteristic,
    pub Length: Characteristic,
}

pub fn avg_or_none(total: Option<i32>, n: Option<i32>) -> Option<f64> {
    if total.is_none() || n.is_none() {
        return None;
    }
    Some(total.unwrap() as f64 / n.unwrap() as f64)
}
